c8d: Adjust "image list" to return only a single item for each image store entry

tianon opened the pull request:
This will return a single entry for each name/value pair, and for now all the "image specific" metadata (labels, config, size) should be either "default platform" or "first platform we have locally" (which then matches the logic for commands like `docker image inspect`, etc) with everything else (just ID, maybe?) coming from the manifest list/index.  That leaves room for the longer-term implementation to add new fields to describe the _other_ images that are part of the manifest list/index.

Refs #45948, #45735 (especially https://github.com/moby/moby/pull/45735#issuecomment-1593670989)

rumpl commented on daemon/containerd/image_list.go:
```diff
@@ -79,51 +82,61 @@ func (i *ImageService) Images(ctx context.Context, opts types.ImageListOptions)
 			return nil, err
 		}
+		if len(presentImages) == 0 {
+			// TODO we should probably show *something* for images we've pulled but are 100% shallow or an empty manifest list/index ("tianon/scratch:index" is an empty example image index and "tianon/scratch:list" is an empty example manifest list)
```
Unrelated to listing images, should we even pull these images? Or rather shouldn't we fail and delete images like these? Since we know we can't run them.

thaJeztah commented:
I do think that the current situation is confusing, and that this PR brings us one step closer to the intended outcome;

```bash
docker pull busybox:latest
1fa89c01cd04: Already exists
fc9db2894f4e: Already exists
3fbc63216742: Already exists
docker.io/library/busybox:latest

docker image ls
REPOSITORY         TAG               IMAGE ID       CREATED         SIZE
busybox            latest            3fbc63216742   5 seconds ago   6.09MB

docker image inspect --format '{{.Architecture}}' 3fbc63216742
arm64

docker image inspect --format '{{.Size}}' 3fbc63216742
1920927

docker pull --platform=linux/amd64 busybox:latest
3fbc63216742: Already exists
023917ec6a88: Download complete
a416a98b71e2: Download complete
3f4d90098f5b: Download complete
docker.io/library/busybox:latest

docker image ls
REPOSITORY         TAG               IMAGE ID       CREATED         SIZE
busybox            latest            3fbc63216742   5 seconds ago   6.09MB
busybox            latest            3fbc63216742   9 days ago      6.61MB

docker image inspect --format '{{.Architecture}}' 3fbc63216742
arm64

docker image inspect --format '{{.Size}}' 3fbc63216742
1920927
```


Even though there's 2 images shown, there's no way to interact with them individually; removing `busybox:latest` will remove both, and so does removing an image by "ID".

It currently also breaks Desktop's Dashboard; here's the overview of images (the overview does not show the right size for each):

<img width="1368" alt="Screenshot 2023-07-20 at 10 16 29" src="https://github.com/moby/moby/assets/1804568/92b6405c-66e5-455e-81ae-dc04e48d7132">


Search for "busybox", and you now have 4 images:

<img width="1348" alt="Screenshot 2023-07-20 at 10 16 37" src="https://github.com/moby/moby/assets/1804568/cb70b429-e8c7-47d5-8378-e55eb3dac30b">


Search again, and now you have 7 images:

<img width="1349" alt="Screenshot 2023-07-20 at 10 16 49" src="https://github.com/moby/moby/assets/1804568/902fa476-91c1-41c7-80a9-2873f9af65c1">

rumpl commented:
> Even though there's 2 images shown, there's no way to interact with them individually

Exactly why my proposal was to _temporarily_ show the manifest ID and not the manifest list ID. Also adding the [platform to the images list would be a nice way to tell them appart](https://github.com/moby/moby/pull/45735)

> I do think that the current situation is confusing, and that this PR brings us one step closer to the intended outcome;

I agree that it's confusing currently, but I don't agree that this takes us any closer to the end goal. This change is also confusing, building a multi-platofrm image yields only one image in `docker images`, which is very confusing. That's why I think we should hold on and work on the final multi-platform aware API

thaJeztah commented:
>  building a multi-platofrm image yields only one image in docker images, which is very confusing.

That's why it's a multi-platform image. If you want multiple platform images, you would give them a different name.

vvoland committed:
c8d: Adjust "image list" to return only a single item for each image store entry

This will return a single entry for each name/value pair, and for now
all the "image specific" metadata (labels, config, size) should be
either "default platform" or "first platform we have locally" (which
then matches the logic for commands like `docker image inspect`, etc)
with everything else (just ID, maybe?) coming from the manifest
list/index.

That leaves room for the longer-term implementation to add new fields to
describe the _other_ images that are part of the manifest list/index.

Co-authored-by: Tianon Gravi <admwiggin@gmail.com>

Signed-off-by: Paweł Gronowski <pawel.gronowski@docker.com>

thaJeztah commented on daemon/containerd/image_list.go:
```diff
@@ -174,26 +186,60 @@ func (i *ImageService) Images(ctx context.Context, opts imagetypes.ListOptions)
+		}
+
+		if len(presentImages) == 0 {
+			// TODO we should probably show *something* for images we've pulled
+			// but are 100% shallow or an empty manifest list/index
+			// ("tianon/scratch:index" is an empty example image index and
+			// "tianon/scratch:list" is an empty example manifest list)
```
Ah, yes, was about to ask if we could test this. As these are minimal images, perhaps we could include them in our "frozen images"?

I'm fine with doing that in a follow-up if it's complicated though.

vvoland replied:
Not sure if it makes sense to handle this right now.

1. It's impossible to pull such image, you can only load it via `docker load`.
2. Even if it is loaded, there's nothing that can be done with that image (except removing/pushing it).

thaJeztah reviewed (approved):
LGTM

vvoland commented:
Good call, I added test and combined the image size.
It turned up to be a bit more code than I expected: https://github.com/moby/moby/pull/47449 🙈 

Do we want to merge this as-is and then review the follow up separately?
